src/shared_data.cc

```diff
@@ -56,6 +56,43 @@ WasmResult SharedData::get(std::string_view vm_id, const std::string_view key,
   return WasmResult::NotFound;
 }
 
+WasmResult SharedData::keys(std::string_view vm_id, const std::string_view key_prefix,
```

Comment:
The `key_prefix` is rather unusual for `keys()` method. What's the reason for including it here parameter? Is this a workaround for a single KV store in current version of Proxy-Wasm or is there another reason?

Reply:
Since SharedData is shared by all plugins within a given VM ID, the guidance is to prefix your keys with a value specific to your plugin (I've seen this written somewhere, but I can't find it to reference right now).  By offering this in the API, we simplify the job of the plugin writer, and encourage a best-practice of only working with their own data-set.

Without this, the first thing the plugins will have to do is filter out any keys that aren't relevant to them (or, worst case, they remove data belonging to another plugin).

Reply:
Keep memory usage as low as possible was our motivation.    That said, looking at some graphs right now, it looks like we see significant additional virtual memory usage as the number of WASM VMs grow (~10G/VM !?), we're not actually seeing all that much real memory usage.

The docs don't give much detail on resource utilization/behavior -- so perhaps my assumption here was wrong.

What are your expectations for increase in memory utilization per VM?  (We're using V8 right now, for what it's worth).

Reply:
WasmVMs cannot load multiple Wasm modules (ignoring module linking), similarly to how you cannot load multiple binaries in the same process. As such, VM ID is only a hint, and multiple instances of the same Proxy-Wasm plugin refer to the same WasmVM only if they have the same VM ID and plugin configuration. Different Proxy-Wasm plugins are always loaded in different WasmVMs, even if they provide the same VM ID.

As such, my question about the key prefix remains - do you have any reason to reuse the same VM ID across different plugins? Right now, it seems like we're adding the key prefix as a workaround for the poor default (VM ID sharing), which could be solved by simply using different VM IDs for different plugins.

As for the memory usage, I didn't look at it in a while, but 10 GiB VSZ per WasmVM sounds about right with V8, and RSS should be around 1 MiB per WasmVM (+ fixed V8 overhead that should be around 10 MiB per process).

V8 uses virtual memory for 2 reasons:
- Guard pages, which allow using signal handlers instead of explicit bound checks, which results in _significant_ performance gain (see: #144).
- Pointer compression, which reduces overall memory consumption, and improves performance (although I'm not sure if and how much does it affect Wasm), see: https://v8.dev/blog/pointer-compression.

Reply:
I hadn't appreciated that the plugin configuration was used to determine whether a distinct VM was created.  With that in mind, I don't think the key prefix is necessary, no.

I'll update the PR to remove it. Thanks for the info!

Reply:
Sorry, I misspoke. All instances of the same Proxy-Wasm plugin with the same VM ID and VM configuration will be attached to the same WasmVM, regardless of the plugin configuration.

Basically, WasmVM's key is derived from: plugin's bytecode, VM ID and VM configuration.

Let me know if that changes anything wrt key prefix.